plugins/Model_GAN128/Trainer.py

```diff
@@ -0,0 +1,222 @@
+        return distorted_input, fake_output, fake_output64, alpha, fn_generate, fn_mask, fn_abgr, fn_bgr
+
+    def define_loss(self, netD, real, fake_argb, fake_sz64, distorted, vggface_feat=None):   
+        alpha = Lambda(lambda x: x[:,:,:, :1])(fake_argb)
+        fake_rgb = Lambda(lambda x: x[:,:,:, 1:])(fake_argb)
+        fake = alpha * fake_rgb + (1-alpha) * distorted
+        
+        if self.use_mixup:
+            dist = Beta(self.mixup_alpha, self.mixup_alpha)
+            lam = dist.sample()
+            # ==========
+            mixup = lam * concatenate([real, distorted]) + (1 - lam) * concatenate([fake, distorted])
+            # ==========
+            output_mixup = netD(mixup)
+            loss_D = self.loss_fn(output_mixup, lam * K.ones_like(output_mixup)) 
+            #output_fake = netD(concatenate([fake, distorted])) # dummy
+            loss_G = 1 * self.loss_fn(output_mixup, (1 - lam) * K.ones_like(output_mixup))
+        else:
+            output_real = netD(concatenate([real, distorted])) # positive sample
+            output_fake = netD(concatenate([fake, distorted])) # negative sample   
+            loss_D_real = self.loss_fn(output_real, K.ones_like(output_real))    
+            loss_D_fake = self.loss_fn(output_fake, K.zeros_like(output_fake))   
+            loss_D = loss_D_real + loss_D_fake
+            loss_G = 1 * self.loss_fn(output_fake, K.ones_like(output_fake))  
+        # ==========  
+        loss_G += K.mean(K.abs(fake_rgb - real))
+        loss_G += K.mean(K.abs(fake_sz64 - tf.image.resize_images(real, [64, 64])))
+        # ==========
+        
+        # Perceptual Loss
+        if not vggface_feat is None:
+            def preprocess_vggface(x):
+                x = (x + 1)/2 * 255 # channel order: BGR
+                x -= [93.5940, 104.7624, 129.]
+                return x
+            pl_params = (0.02, 0.3, 0.5)
+            real_sz224 = tf.image.resize_images(real, [224, 224])
+            real_sz224 = Lambda(preprocess_vggface)(real_sz224)
+            # ==========
+            fake_sz224 = tf.image.resize_images(fake_rgb, [224, 224]) 
+            fake_sz224 = Lambda(preprocess_vggface)(fake_sz224)
+            # ==========   
+            real_feat55, real_feat28, real_feat7 = vggface_feat(real_sz224)
+            fake_feat55, fake_feat28, fake_feat7  = vggface_feat(fake_sz224)    
+            loss_G += pl_params[0] * K.mean(K.abs(fake_feat7 - real_feat7))
+            loss_G += pl_params[1] * K.mean(K.abs(fake_feat28 - real_feat28))
+            loss_G += pl_params[2] * K.mean(K.abs(fake_feat55 - real_feat55))
+        
+        return loss_D, loss_G
+
+    def show_sample(self, display_fn):
+        _, wA, tA = next(self.train_batchA)
+        _, wB, tB = next(self.train_batchB)
+        display_fn(self.showG(tA, tB, self.path_A, self.path_B), "raw")
+        display_fn(self.showG(wA, wB, self.path_bgr_A, self.path_bgr_B), "masked")

```

Comment:
The masked image preview is quite warped. Should this be passing `tA, tB` instead of `wA, wB`? Commit https://github.com/deepfakes/faceswap/commit/aa063975021f18fb5a7f2e7b824dca21d5276008 from @oatssss fixes this on his GAN64 v2 branch.

Related, https://github.com/deepfakes/faceswap-playground/issues/65

Reply:
@shaoanlu what information does the "raw" preview give us? "raw" and "masked" both seem to give a masked output, but I believe "raw" is manually computed using the output rgb and alpha channels.